scipy/signal/_signaltools.py

```diff
@@ -3559,20 +3562,28 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=False):
     0.06  # random
 
     """
+    xp = array_namespace(data)
     if type not in ['linear', 'l', 'constant', 'c']:
         raise ValueError("Trend type must be 'linear' or 'constant'.")
-    data = np.asarray(data)
-    dtype = data.dtype.char
-    if dtype not in 'dfDF':
-        dtype = 'd'
+    data = xp.asarray(data)
+    dtype = data.dtype
+    if data.dtype not in [xp.float32, xp.float64, xp.complex128, xp.complex64]:
+        dtype = xp.float64
     if type in ['constant', 'c']:
-        ret = data - np.mean(data, axis, keepdims=True)
+        ret = data - xp.mean(xp.asarray(data, dtype=dtype), axis=axis, keepdims=True)
         return ret
     else:
         dshape = data.shape
         N = dshape[axis]
-        bp = np.sort(np.unique(np.concatenate(np.atleast_1d(0, bp, N))))
-        if np.any(bp > N):
+        if isinstance(bp, int):
+            bp = xp.sort(xp.unique(xp.asarray([0, bp, N])))
+        else:
+            bp = xp.asarray(bp)
+            new_bp = xp.empty(size(bp) + 1)
+            new_bp[:size(bp)] = bp
+            new_bp[size(bp)] = N
+            bp = xp.sort(xp.unique(xp.asarray([0] + new_bp)))
+        if xp.any(bp > N):
```

Comment:
We sorted `bp` already so you can compare the last element instead to save some pennies. 

Error message is also a bit cryptic though not relevant.